### PR TITLE
SSZ enums validation

### PR DESCRIPTION
**Pull request by zah**
This is a more proper fix replacing #1208

### CODE REVIEW

**Comment by arnetheduck** [comment]
I generally feel that using the `enum` type in networking protocol/serialization definitions at large is wrong - there's a semantic mismatch between what the language wants (how nim defines `enum`) and how the world of serialization works / what is good practise.

in the wild, "out-of-range" enum values are the norm for distributed systems so trying to force `enum` into that bucked sends the wrong signal to whoever is designing a protocol or serialization format - they're simply a bad idea in the context - using const values is generally the better option - specifically because of the hole issues, the start of the range etc

it's merely a coincidence that sometimes, error values like the one we encountered are serially numbered starting from 0 or 1 - more often, the values in protocols are just "random" tags and not enumerations really.

**Comment by arnetheduck** [comment]
In fact, this PR is a good example of how the semantic mismatch leads to bugs: https://github.com/ethereum/eth2.0-specs/blob/v0.12.0/specs/phase0/p2p-interface.md#goodbye says:

> Clients MAY use reason codes above 128 to indicate alternative, erroneous request-specific responses.

an enum does not capture this correctly, and anyone reading the `goodbye` declaration in this code will think that values outside of the box should be treated as if the client was sending an invalid ssz value.

**Comment by zah** [comment]
SSZ doesn't have any provisions for forward or backward compatibility and half-hearted attempts to introduce it somewhere are not much of an improvement. Sending an invalid enum value is a violation of the protocol and after this patch, it will be reported as such. The code used to fail with a `Defect`, now it's going to report a regular recoverable error.

File: beacon_chain/ssz/bytes_reader.nim [hunk 1]
```diff
@@ -57,8 +57,9 @@ func fromSszBytes*(T: type Version, bytes: openarray[byte]): T {.raisesssz.} =
     raiseIncorrectSize T
   copyMem(result.addr, unsafeAddr bytes[0], sizeof(result))
 
-template fromSszBytes*(T: type enum, bytes: openarray[byte]): auto  =
-  T fromSszBytes(uint64, bytes)
+func fromSszBytes*(T: type enum, bytes: openarray[byte]): T {.raisesssz.} =
```
Comment (arnetheduck) [inline beacon_chain/ssz/bytes_reader.nim:60]: this function should be removed completely actually

**Comment by arnetheduck** [comment]
In the `goodbye` case, the message is defined as taking an `uint64`, not a `nim enum` - the concept of nim enums doesn't have a corresponding construct in the SSZ spec, so we should not be making one up if the mapping is not perfect, semantically.

**Comment by zah** [comment]
Hrm, alright, I spoke before reading the latest spec for `goodbye`.

**Pull request closed without merging**